### ISSUE TITLE
Contenteditable - replace to insert html.

### ISSUE THREAD

**Issue by andrejkaurin** [yuku-t/jquery-textcomplete#105]
How about having replace function to return something like this

replace: function (word) {
            return '&lt;span style="background: #BBBBBB;color: #FFFFFF;"&gt;' + word + '&lt;/span&gt;'
        }

It currently inserts as code and not as html element into content editable.

**Comment by calvinalvin** [comment]
I found that using '&nbsp;' for the dummy innerHtml gives better results across browsers:

```
      if (this.el.isContentEditable) {
        range.selectNodeContents(range.startContainer);
        range.deleteContents();

        var dummy = document.createElement('span');
        dummy.innerHTML = '&nbsp;';
        range.insertNode(dummy);

        var node = range.createContextualFragment(pre + post);
        range.insertNode(node);
        range.setStartAfter(dummy);
        range.collapse(true);

        sel.removeAllRanges();
        sel.addRange(range);
      } else {
        this.$el.val(pre + post);
        this.el.selectionStart = this.el.selectionEnd = pre.length;
      }
```

**Comment by surajhell88** [comment]
Try this which worked for me, referred http://stackoverflow.com/questions/6690752/insert-html-at-caret-in-a-contenteditable-div

        if (this.el.contentEditable == 'true') {
          range.selectNodeContents(range.startContainer);
          range.deleteContents();
          var el = document.createElement("div");
          el.innerHTML = pre + post;
          var frag = document.createDocumentFragment(), node, lastNode;
          while ( (node = el.firstChild) ) {
              lastNode = frag.appendChild(node);
          }
          range.insertNode(frag);

          // Preserve the selection
          if (lastNode) {
              range = range.cloneRange();
              range.setStartAfter(lastNode);
              range.collapse(true);
              sel.removeAllRanges();
              sel.addRange(range);
          }
        } else {
          this.$el.val(pre + post);
          this.el.selectionStart = this.el.selectionEnd = pre.length; 
        }

**Comment by hclarke3** [comment]
While the solutions above worked, they also allow the ability to execute html inline to the input area. So, for instance, you could type `<h1>Hello</h1>` and upon select a new header element would be inserted into the contenteditable element. I think I've managed to fix all of the hiccups with the function below if anyone is interested:

```javascript
// allow insertion of html elements
if (this.el.isContentEditable) {
  range.selectNodeContents(range.startContainer);
  range.deleteContents();

  var dummy = document.createElement('span');
  dummy.innerHTML = '&nbsp;';
  range.insertNode(dummy);

  var fullText = pre + post;
  var elIdx = fullText.indexOf(newSubstr);

  // insertNode places nodes at the beginning of the range,
  // so insert everything after the new node first
  var postElTextNode = document.createTextNode(fullText.substr(elIdx + fullText.length + 1));
  range.insertNode(postElTextNode);

  // insert the new node
  var node = range.createContextualFragment(newSubstr);
  range.insertNode(node);

  // insert everything before the new node
  var preElTextNode = document.createTextNode(fullText.substr(0, elIdx));
  range.insertNode(preElTextNode);

  range.setStartAfter(dummy);
  range.collapse(true);

  sel.removeAllRanges();
  sel.addRange(range);
} else {
    this.$el.val(pre + post);
    this.el.selectionStart = this.el.selectionEnd = pre.length;
}
```

**Comment by yuku-t** [comment]
fixed by #217
